Implement SDL-0230: SPP resource management for Android.

shiniwat opened the pull request:
Fixes #1132 

This PR is **ready** for review, but localization task still remains.

### Risk
This PR makes **no** API changes.

### Testing Plan
This PR includes unit test in MultiplexTransportTest, but a part of this change is in error UX, which may need additional test.

### Summary
This change is to implement SDL-0230: SPP resource management for Android.

### Changelog
##### Breaking Changes
* N/A

##### Enhancements
* N/A

##### Bug Fixes
* N/A

### Tasks Remaining:
- [x] Localization task

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

shiniwat commented:
While this PR does not include localization part, the implementation of SDL-0230 is ready for review. Thanks.

bilal-alsharifi commented:
@shiniwat Could you please include more details about how to test this PR?

shiniwat commented:
Hi @bilal-alsharifi,
The PR includes testSetState, but MultiplexTransportTests might need other Unit Tests in either MultiplexTransportTest or MultiplexBluetoothTransportTest (or even both). They both look like missing many methods, however.
If you can specify what tests are needed for this PR, I'm happy to add them.

bilal-alsharifi commented:
Thanks, @shiniwat. Sorry, my question was not very clear. I meant to ask if there is an easy way to make the SPP channel runs out of available resources on the phone. I just would like to see the case happen in action in order to smoke test the PR.

shiniwat commented:
OK, I have created a sample app for this exact purpose at https://github.com/shiniwat/consumespp

The app (called ConsumeSPPServices) consumes all available SPP resources, so that Android system will run into SPP resource shortage situation.

bilal-alsharifi commented:
Thanks, @shiniwat  that sample app will be very helpful for testing.

shiniwat commented on android/sdl_android/src/main/java/com/smartdevicelink/transport/SdlRouterService.java:
```diff
@@ -3643,5 +3651,55 @@ public void clear(){
+	 * notifySppError: utilize notification channel to notify the SPP out-of-resource error.
+	 */
+	@TargetApi(11)
+	public void notifySppError() {
```
Right, it doesn't have to be public.

shiniwat commented on android/sdl_android/src/main/java/com/smartdevicelink/transport/SdlRouterService.java:
```diff
@@ -3643,5 +3651,55 @@ public void clear(){
+	/**
+	 * notifySppError: utilize notification channel to notify the SPP out-of-resource error.
+	 */
+	@TargetApi(11)
```
Make sense, thanks.

shiniwat commented on android/sdl_android/src/main/java/com/smartdevicelink/transport/SdlRouterService.java:
```diff
@@ -3643,5 +3651,55 @@ public void clear(){
+		builder.setOngoing(false);
+
+		Log.e(TAG, "Notification: notifySppError entering");
+		synchronized (NOTIFICATION_LOCK) {
```
I thought we need the NOTIFICATION_LOCK whenever NotificationManager.notify gets called. If the lock is specific to Foreground notification, the lock should be renamed something like "FOREGROUND_NOTIFICATION_LOCK"?
I will double check if the lock is indeed required or not...

bilal-alsharifi replied:
Yes, the lock is not needed for the SPP error notification. I actually like the idea of renaming `NOTIFICATION_LOCK` to `FOREGROUND_NOTIFICATION_LOCK` to make things more clear. So feel free to rename that as part of the PR.

shiniwat replied:
Thanks for the suggestion. My commit at b91fc55 renamed the lock as you suggested.

shiniwat commented on android/sdl_android/src/androidTest/java/com/smartdevicelink/transport/MultiplexTransportTest.java:
```diff
@@ -80,4 +85,19 @@ public void testThreadRecoverability(){
+		Bundle bundle = new Bundle();
+		bundle.putByte(MultiplexBaseTransport.ERROR_REASON_KEY, MultiplexBaseTransport.REASON_SPP_ERROR);
+		btTransport.setState(MultiplexBaseTransport.STATE_ERROR, bundle);
+		assertTrue(true); // succeeded
```
The suggestion makes a lot sense, thanks.

shiniwat commented:
Hi @bilal-alsharifi, thanks for the review.
I have updated the PR based on your suggestions.

bilal-alsharifi commented:
Thanks, @shiniwat for making the changes. I commented on the notification lock one. The other changes look good.

bilal-alsharifi commented on android/sdl_android/src/main/res/values/strings.xml:
```diff
@@ -4,4 +4,7 @@
     <string name="default_lockscreen_warning_message">Swipe down to dismiss, acknowledging that you are not the driver.</string>
+    <string name="spp_out_of_resource">Too many apps are using Bluetooth</string>
+    <string name="notification_title">SmartDeviceLink</string>
+    <string name="sdl_error_notification_channel_name">SDL Error</string>
```
Could you please add translations to these strings in all the languages that are specified in the proposal?

shiniwat replied:
Hi @bilal-alsharifi,
The commit b626fa6 added localized strings for all 29 languages defined in MOBILE_API.
Please review. Thanks.

bilal-alsharifi replied:
Thanks, @shiniwat for adding the languages. Can we also add strings for different dialects for the same language? like
English: en_US, en_AU, en_GB, en_IN, en_SA
Spanish: es_ES, es_MX
French: fr_CA, fr_FR
Portuguese: pt_BR, pt_PT

As these variations also exist in the `MOBILE_API `

shiniwat replied:
Well, actually adding non-locale specific strings (i.e. language only) for most languages is intentional. Most languages do not differentiate short messages per locale. For instance, the message "Too many apps are using Bluetooth" should be the same for English in US, England, Australia, India, etc. French message should be the same for Canada and France, too. So I basically did not add the string per locale if the strings are the same. Obvious exception is Chinese -- Simplified Chinese and Traditional Chinese are obviously different, so I added both.

bilal-alsharifi replied:
That actually makes all since. My only concern is that it is mentioned in the accepted proposal that there should be strings for all variations including different dialects for the same language. 
Could you please make a quick proposal revision just to clarify the ["Localization of SPP error UX"](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0230-spp-resource-management-for-android.md#localization-of-spp-error-ux) section as you did in the last comment to make sure that the SDLC agrees? For more info on revising a proposal please check [this page](https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md#revising-previously-accepted-proposals).

shiniwat replied:
As I re-read the "[Localization of SPP error UX](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0230-spp-resource-management-for-android.md#localization-of-spp-error-ux)" section, the section refers to a list of language enums from[ MOBILE_API](https://github.com/smartdevicelink/rpc_spec/blob/master/MOBILE_API.xml), which needs to be supported.
However, the section does not mention about implementation details. As far as we can support all language enums that [MOBILE_API](https://github.com/smartdevicelink/rpc_spec/blob/master/MOBILE_API.xml) refers to, we should be OK.
Adding locale-specific language (e.g. Simplified Chinese) or language only (e.g. English for all English-speaking locales) resource depends on how Android resource management works. Therefore, it's my understanding that current commit still follows the latest revision of the proposal and no need to revise.

bilal-alsharifi replied:
Okay, thanks for clarifying.

bilal-alsharifi commented:
Livio will provide an icon for error notification that can be added to the PR and voted on by the SDLC as well.

bilal-alsharifi commented:
@shiniwat I tested the PR again after you merged in the latest `develop` and realized that the notification stopped getting displayed at all. 
I did some debugging and found that the issue is that we call this API to cancel all notifications in `exitForeground()` in the `SdlRouterService` class:
```
notificationManager.cancelAll();
```
That was okay before adding the new spp notification. But now we don't want to cancel them all. We should replace the previous call with the follwing to cancel only the foreground service notification:
```
notificationManager.cancel(FOREGROUND_SERVICE_ID);
```

bilal-alsharifi commented:
@shiniwat we made an icon for the SPP error notification. You can set that in:
```
builder.setLargeIcon(icon)
```
We can later ask the SDLC to vote on it.
[res.zip](https://github.com/smartdevicelink/sdl_java_suite/files/4636336/res.zip)

shiniwat commented:
>```
> notificationManager.cancel(FOREGROUND_SERVICE_ID);
> ```
@bilal-alsharifi, thanks for catching this. In my testing, SPP error notification was not canceled anyhow, but your suggested fix indeed makes sense. I will update the PR accordingly.

>@shiniwat we made an icon for the SPP error notification. You can set that in:

Thanks for providing the icon as well. I will update it as well.

shiniwat commented:
Hi @bilal-alsharifi, I have updated the cancel issue and icon resource in accordance with your feedback. Thanks.

bilal-alsharifi commented:
Thanks, @shiniwat for making the changes. I will take a look.

>@bilal-alsharifi, thanks for catching this. In my testing, SPP error notification was not canceled anyhow, but your suggested fix indeed makes sense. I will update the PR accordingly.

You are right, the `cancelAll()` call does nothing on some devices for some reason but it cancels all notifications on others.

bilal-alsharifi commented on android/sdl_android/src/main/java/com/smartdevicelink/transport/SdlRouterService.java:
```diff
@@ -3707,5 +3714,55 @@ public void clear(){
+		builder.setContentText(getString(R.string.spp_out_of_resource));
+
+		//We should use icon from library resources if available
+		// @REVIEW: large icon uses SPP error specific icon, but can we reuse the small icon with the default tray icon?
```
Keeping the current small icon should be fine. Can we remove this comment?

shiniwat replied:
I just removed the @REVIEW comment at aac3702 commit. Thanks.

bilal-alsharifi commented:
@theresalech The PR has been updated to include the localization strings in [this commit](https://github.com/smartdevicelink/sdl_java_suite/pull/1144/commits/b626fa6f55d7ddb3384d0f128e1b98c9f9bd073c) and the icon in [this commit](https://github.com/smartdevicelink/sdl_java_suite/pull/1144/commits/6c0eea6498cc7d4cfd33333ef346cd95398423b1). Can we ask the SDLC to vote on them?

theresalech commented:
@bilal-alsharifi, yes I will ask for SDLC Steering Committee vote on these items during the 2020-05-26 meeting.

shiniwat commented:
Attached screenshots to demonstrate how SPP error notification looks in some languages.
![SPP_error_screenshots](https://user-images.githubusercontent.com/25340298/82526391-e215aa80-9b6e-11ea-9e62-8fc67758c71f.png).
Hope this helps for upcoming review (for 2020-05-26 steering committee meeting).

shiniwat commented:
Hi @bilal-alsharifi 
I double checked all languages listed in MOBILE_API by actually running hello_sdl_android on my Pixel, and found three issues:
1. Dutch resource was missing. Added values-nl/strings.xml to fix.
2. Indonesian resource should be values-in; values-id was wrong.
3. On my Pixel, values-he did not work for Hebrew. values-iw worked instead. Added both eventually.
Those three fixes are in commit at https://github.com/XevoInc/sdl_android/commit/270a0f05da70b92b805bb31501710aa38cb842cb

shiniwat commented:
FYI, but here are screenshots for all other languages (than previously shown ones):
![SPP_error_screenshots2](https://user-images.githubusercontent.com/25340298/82651333-f7fe9a80-9c56-11ea-9930-bfa9ae92f2ae.png)

![SPP_error_screenshots3](https://user-images.githubusercontent.com/25340298/82651354-05b42000-9c57-11ea-95eb-b0824ebef67b.png)

![SPP_error_screenshots4](https://user-images.githubusercontent.com/25340298/82651386-0ea4f180-9c57-11ea-8b1f-67e6d21aacf0.png)

theresalech commented:
@bilal-alsharifi the SDLC Steering Committee approved the localization strings and icon used in this PR during the 2020-05-26 Steering Committee meeting.

bilal-alsharifi commented:
@shiniwat There is one merge conflict with develop. Could you please resolve that before merging the PR?

shiniwat commented:
Hi @bilal-alsharifi,
I have merged current develop branch and resolved the conflict in SdlRouterService. Sorry for the inconvenience.

bilal-alsharifi commented:
Thanks, @shiniwat for fixing the merge conflict.
